### ISSUE TITLE
False negative when using magic methods

### ISSUE THREAD

**Issue by fsouza** [python/mypy#8027]
Note: if you are reporting a wrong signature of a function or a class in
the standard library, then the typeshed tracker is better suited
for this report: https://github.com/python/typeshed/issues

Please provide more information to help us understand the issue:

* Are you reporting a bug, or opening a feature request?

Bug?

* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

```python
from typing import List
from typing import Optional


class CustomClass:
    def __getitem__(self, key: str) -> Optional[str]:
        pass


items: List[CustomClass] = []
[item["a"].strip() for item in items if item["a"] is not None]
```

The one below doesn't type check, as expected:

<details>

```python
from typing import List
from typing import Optional


class CustomClass:
    def getitem(self, key: str) -> Optional[str]:
        pass


items: List[CustomClass] = []
[item.getitem("a").strip() for item in items if item.getitem("a") is not None]
```

</details>

* What is the actual behavior/output?

```
Success: no issues found in 1 source file
```

* What is the behavior/output you expect?

It should not type check, as repeated calls to __getitem__ are not guaranteed to return the same value.

* What are the versions of mypy and Python you are using?
  Do you see the same issue after installing mypy from Git master?

```
％ mypy --version
mypy 0.750+dev.1013b63eac2bef395952f2784fb3390dc181dec7
％ python --version
Python 3.7.5
```

* What are the mypy flags you are using? (For example --strict-optional)

``--strict --scripts-are-modules``

**Comment by ilevkivskyi** [comment]
This is a duplicate of https://github.com/python/mypy/issues/7339

**Comment by fsouza** [comment]
@ilevkivskyi doesn't #7339 cover a false positive? This here is a false negative 🤔 Or is that issue supposed to cover everything-indexing?

Type refinement can't be safely accomplished in the sample code, therefore mypy should reject it.

**Comment by ilevkivskyi** [comment]
OK, I see. Then this is a duplicate of a different issue https://github.com/python/mypy/issues/4168 (modulo attribute -> index). The latter has also false-positive label because overly narrow types cause troubles with `--strict-equality`

**Comment by fsouza** [comment]
Thanks for the info!
